### PR TITLE
[feature] Add NormalisedTK2Predicate & NormaliseTK2 pass

### PR DESCRIPTION
**Pull request by lmondada**

### CODE REVIEW

File: tket/tests/test_TwoQubitCanonical.cpp [hunk 1]
```diff
@@ -540,12 +543,15 @@ static void check_decompose_2cx_VD(const Eigen::Matrix4cd &U) {
 static void check_decompose_2cx_DV(const Eigen::Matrix4cd &U) {
   auto [circ, z0] = decompose_2cx_DV(U);
   unsigned n_cx = 0;
+  static const std::set<OpType> expected_1q_gates = {
```
Comment (cqc-alec) [inline tket/tests/test_TwoQubitCanonical.cpp:546]: Update docstring of `decompose_2cx_DV`.

File: tket/tests/test_TwoQubitCanonical.cpp [hunk 2]
```diff
@@ -540,12 +543,15 @@ static void check_decompose_2cx_VD(const Eigen::Matrix4cd &U) {
...
   for (const Command &cmd : circ) {
     OpType optype = cmd.get_op_ptr()->get_type();
     if (optype == OpType::CX) {
       n_cx++;
     } else {
-      CHECK(optype == OpType::TK1);
+      CHECK(expected_1q_gates.contains(optype));
     }
   }
   CHECK(n_cx <= 2);
```
Comment (cqc-alec) [inline tket/tests/test_TwoQubitCanonical.cpp:557]: Could also add basic serialization tests in test_json.cpp.

**Review by cqc-alec — CHANGES_REQUESTED** [review]
Nice, just a few comments.

File: tket/src/Circuit/CircPool.cpp [hunk 3]
```diff
@@ -1034,6 +1034,144 @@ Circuit NPhasedX_using_PhasedX(
...
+  }
+  if (c && *c > .5) {
+    ec -= 1;
+    *c -= 1.;
+    post.add_op<unsigned>(OpType::Z, {0});
+    post.add_op<unsigned>(OpType::Z, {1});
+    pre.add_phase(-0.5);
+  }
+  // Cheeky way of reversing order of ops.
+  post = post.dagger();
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Circuit/CircPool.cpp:1164]: Yes... all gates added are self-adjoint, up to `S/Sdg` and `V/Vdg`, but that is handled in the `auto conj` lambda (see comment line 1050).

File: pytket/binders/passes.cpp [hunk 4]
```diff
@@ -369,11 +368,25 @@ PYBIND11_MODULE(passes, m) {
       "noise into consideration.\n\n"
       "If no fidelities are provided, the TK2 gates will be decomposed "
       "exactly using CX gates.\n\n"
-      "All TK2 gate parameters must be normalised to the Weyl chamber."
+      "All TK2 gate parameters must be normalised, i.e. they must satisfy "
+      "`NormalisedTK2Predicate`."
```
[the remark this answers is not in the document]
Comment (lmondada) [reply pytket/binders/passes.cpp:372]: Done.

File: tket/tests/test_TwoQubitCanonical.cpp [hunk 5]
```diff
@@ -517,12 +517,15 @@ SCENARIO("Test qubit reversal") {
...
+  static const std::set<OpType> expected_1q_gates = {
+      OpType::TK1, OpType::H, OpType::V, OpType::Vdg, OpType::S,
+      OpType::Sdg, OpType::X, OpType::Y, OpType::Z};
   for (const Command &cmd : circ) {
     OpType optype = cmd.get_op_ptr()->get_type();
     if (optype == OpType::CX) {
       n_cx++;
     } else {
-      CHECK(optype == OpType::TK1);
+      CHECK(expected_1q_gates.contains(optype));
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/tests/test_TwoQubitCanonical.cpp:528]: Done.

File: tket/tests/test_TwoQubitCanonical.cpp (hunk 1 above)
Comment (lmondada) [reply tket/tests/test_TwoQubitCanonical.cpp:546]: Done.

File: tket/tests/test_TwoQubitCanonical.cpp (hunk 2 above)
Comment (lmondada) [reply tket/tests/test_TwoQubitCanonical.cpp:557]: Done.

File: tket/src/Transformations/BasicOptimisation.cpp [hunk 6]
```diff
@@ -708,6 +709,34 @@ Transform absorb_Rz_NPhasedX() {
...
 }
 
+Transform normalise_TK2() {
+  return Transform([](Circuit &circ) {
+    bool success = false;
+    VertexSet bin;
+
+    BGL_FORALL_VERTICES(v, circ.dag, DAG) {
+      Op_ptr op = circ.get_Op_ptr_from_Vertex(v);
+      if (op->get_type() == OpType::TK2) {
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Transformations/BasicOptimisation.cpp:719]: Done.

File: tket/src/Predicates/Predicates.cpp [hunk 7]
```diff
@@ -656,6 +658,32 @@ std::string GlobalPhasedXPredicate::to_string() const {
   return auto_name(*this);
 }
 
+bool NormalisedTK2Predicate::verify(const Circuit& circ) const {
+  BGL_FORALL_VERTICES(v, circ.dag, DAG) {
+    Op_ptr op = circ.get_Op_ptr_from_Vertex(v);
+    if (op->get_type() == OpType::TK2) {
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Predicates/Predicates.cpp:664]: Done.

**Comment by lmondada** [comment]
Thanks @cqc-alec for your useful review. All your comments should be addressed.

File: tket/src/Transformations/BasicOptimisation.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/src/Transformations/BasicOptimisation.cpp]: Looks like this one is not done, though the other one is?

File: tket/tests/test_TwoQubitCanonical.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/tests/test_TwoQubitCanonical.cpp:528]: Still says
```
 * implemented by the circuit and D = diag(z, z*, z*, z). The circuit C consists
 * of CX and TK1 gates only and has at most 2 CX gates.
```

File: tket/src/Transformations/BasicOptimisation.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/src/Transformations/BasicOptimisation.cpp]: No, my mistake.

File: tket/tests/test_TwoQubitCanonical.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/tests/test_TwoQubitCanonical.cpp:528]: My mistake.

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**